source/extensions/filters/http/cdn_loop/parser.h

```diff
@@ -0,0 +1,299 @@
+#pragma once
+
+#include <iostream>
+
+#include "common/common/statusor.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/string_view.h"
+
+// This file defines a parser for the CDN-Loop header value.
+//
+// RFC 8586 Section 2 defined the CDN-Loop header as:
+//
+//   CDN-Loop  = #cdn-info
+//   cdn-info  = cdn-id *( OWS ";" OWS parameter )
+//   cdn-id    = ( uri-host [ ":" port ] ) / pseudonym
+//   pseudonym = token
+//
+// Each of those productions rely on definitions in RFC 3986, RFC 5234, RFC
+// 7230, and RFC 7231. Their use is noted in the individual parse functions.
+//
+// The parser is a top-down combined parser and lexer that implements just
+// enough of the RFC spec to make it possible count the number of times a
+// particular CDN value appears. The main differences between the RFC's grammar
+// and the parser defined here are:
+//
+// 1. the parser has a more lax interpretation of what's a valid uri-host. See
+//     ParseCdnId for details.
+//
+// 2. the parser allows leading and trailing whitespace around the header
+//     value. See ParseCdnInfoList for details.
+//
+// Each parse function takes as input a ParseContext that tells the
+// function where to start. Parse functions that just need to parse a portion
+// of the CDN-Loop header, but don't need to return a value, should return a
+// ParseContext pointing to the next character to parse. Parse functions that
+// need to return a value should return something that contains a ParseContext.
+//
+// Parse functions that can fail (most of them!) wrap their return value in an
+// Envoy::StatusOr.
+//
+// In the interest of performance, this parser works with string_views and
+// references instead of copying std::strings. The string_view passed into the
+// ParseContext of a parse function must outlive the return value of the
+// function.
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CdnLoop {
+namespace Parser {
+
+// A ParseContext contains the state of the recursive descent parser and some
+// helper methods.
+class ParseContext {
+public:
+  ParseContext(absl::string_view value) : value_(value), next_(0) {}
+  ParseContext(absl::string_view value, absl::string_view::size_type next)
+      : value_(value), next_(next) {}
+
+  // Return true if we have reached the end of value.
```

Comment:
nit: here and the rest, s/Return/Returns

https://google.github.io/styleguide/cppguide.html#Function_Comments

Reply:
Done.